nemo/lightning/megatron_parallel.py

```diff
@@ -945,22 +833,413 @@ def __contains__(self, callback_object) -> bool:
+    """
+
+    pipeline: MegatronParallel[ModelT]
+    data: Union[DataT, Iterator[DataT], List[Iterator[DataT]]]
+    forward_step_func: Callable
+    forward_only: bool
+    micro_batch_size: Optional[int] = None
+    seq_length: Optional[int] = None
+    num_microbatches: Optional[int] = None
+
+    @classmethod
+    def infer(
+        cls,
+        pipeline: MegatronParallel[ModelT],
+        data: DataT,
+        forward_step_func: Callable,
+        forward_only: bool,
+        micro_batch_size: Optional[int] = None,
+        seq_length: Optional[int] = None,
+        num_microbatches: Optional[int] = None,
+    ) -> "MegatronStep[ModelT, DataT]":
+        """
+        Creates a MegatronStep instance, inferring missing parameters if possible.
+
+        This method attempts to infer the micro_batch_size, seq_length, and num_microbatches
+        from the provided data if they are not explicitly specified.
+
+        Args:
+            pipeline (MegatronParallel[ModelT]): The Megatron parallel model pipeline.
+            data (DataT): Input data for the step.
+            forward_step_func (Callable): Function to perform the forward step.
+            forward_only (bool): If True, only perform forward pass (no backward pass).
+            micro_batch_size (Optional[int]): Size of each micro-batch.
+            seq_length (Optional[int]): Sequence length for the current step.
+            num_microbatches (Optional[int]): Number of micro-batches in this step.
+
+        Returns:
+            MegatronStep[ModelT, DataT]: An instance of MegatronStep with inferred parameters.
+        """
+        return cls(
+            pipeline=pipeline,
+            data=data,
+            forward_step_func=forward_step_func,
+            forward_only=forward_only,
+            micro_batch_size=micro_batch_size or cls.infer_micro_batch_size(data),
+            seq_length=seq_length or cls.infer_seq_length(data),
+            num_microbatches=num_microbatches or cls.infer_num_microbatches(data),
+        )
+
+    def __call__(self) -> List[Any]:
+        """
+        Executes the Megatron step.
+
+        This method performs the forward (and optionally backward) pass using the
+        configured forward_backward_func. It ensures all necessary parameters are set
+        before execution.
+
+        Returns:
+            List[Any]: The output of the forward_backward_func, typically containing
+                       loss values and other relevant information.
+
+        Raises:
+            ValueError: If any of num_microbatches, seq_length, or micro_batch_size is not set.
+        """
+        if self.num_microbatches is None:
+            raise ValueError("num_microbatches is not set")
+
+        if self.seq_length is None:
+            raise ValueError("seq_length is not set")
+
+        if self.micro_batch_size is None:
+            raise ValueError("micro_batch_size is not set")
+
+        return self.forward_backward_func(
```

Comment:
Pass `forward_only` into `self.forward_backward_func`.

Reply:
Pushed my fix.

Reply:
Thank you.